### ISSUE TITLE
mongrel2 exits uncleanly

### ISSUE THREAD

**Issue by jkarneges** [fanout/pushpin#47638]
Sometimes Pushpin will log a message like this on exit:

```
[ERR] 2016-10-22 14:09:37.809 m2 http:7999: Exited uncleanly
```

This is a bug in Mongrel2 (a dependency of Pushpin), and not Pushpin itself, but since users may see this error I figured it would be good to have an issue filed until it is fixed. Note that the bug is harmless.

**Comment by spiderpug** [comment]
First of all, thanks for maintaining such a great tool.

I experience a lot of these issues with the provided docker image. Unknown circumstances cause the server/container to fail to start:

```
pushpin_1  | [INFO] 2017-04-05 15:40:29.492 starting...
pushpin_1  | [INFO] 2017-04-05 15:40:29.492 using config: /etc/pushpin/pushpin.conf
pushpin_1  | [INFO] 2017-04-05 15:40:29.524 starting m2 http:7999
pushpin_1  | [ERR] 2017-04-05 15:40:29.532 m2 http:7999: Exited unexpectedly
pushpin_1  | [INFO] 2017-04-05 15:40:29.532 [m2a] starting...
pushpin_1  | [INFO] 2017-04-05 15:40:29.533 [m2a] m2_in connect ipc:///var/run/pushpin/pushpin-m2-out-7999
pushpin_1  | [ERR] 2017-04-05 15:40:29.533 zurl: Exited uncleanly
pushpin_1  | [INFO] 2017-04-05 15:40:29.533 [m2a] m2_out connect ipc:///var/run/pushpin/pushpin-m2-in-7999
pushpin_1  | [INFO] 2017-04-05 15:40:29.533 [m2a] m2_control connect pushpin-m2-7999:ipc:///var/run/pushpin/pushpin-m2-control-7999
pushpin_1  | [ERR] 2017-04-05 15:40:29.534 handler: Exited uncleanly
pushpin_1  | [INFO] 2017-04-05 15:40:29.534 started
pushpin_1  | [INFO] 2017-04-05 15:40:29.533 [m2a] zhttp_in bind ipc:///var/run/pushpin/pushpin-m2zhttp-in
pushpin_1  | [INFO] 2017-04-05 15:40:29.534 [m2a] zhttp_out bind ipc:///var/run/pushpin/pushpin-m2zhttp-out
pushpin_1  | [ERR] 2017-04-05 15:40:29.534 proxy: Exited uncleanly
pushpin_1  | [INFO] 2017-04-05 15:40:29.534 [m2a] zhttp_out_stream bind ipc:///var/run/pushpin/pushpin-m2zhttp-out-stream
pushpin_1  | [INFO] 2017-04-05 15:40:29.534 [m2a] zws_in bind ipc:///var/run/pushpin/pushpin-m2zws-in
pushpin_1  | [INFO] 2017-04-05 15:40:29.535 [m2a] zws_out bind ipc:///var/run/pushpin/pushpin-m2zws-out
pushpin_1  | [INFO] 2017-04-05 15:40:29.535 [m2a] zws_out_stream bind ipc:///var/run/pushpin/pushpin-m2zws-out-stream
pushpin_1  | [INFO] 2017-04-05 15:40:29.535 [m2a] started
pushpin_1  | [INFO] 2017-04-05 15:40:29.535 [m2a] stopping...
pushpin_1  | [INFO] 2017-04-05 15:40:29.535 [m2a] stopped
pushpin_1  | [INFO] 2017-04-05 15:40:29.538 stopped
app_pushpin_1 exited with code 0
```

It only occurs _sometimes_ (I know, those are the best). Unfortunately the server won't boot at all, so it's not that harmless in my case. ;-)

Any idea what might cause this behaviour? Is it somehow related to my Docker Engine (running on OSX here).

**Comment by jkarneges** [comment]
Looks like mongrel2 may be failing to start. Usually this happens if the port is taken by another process, or if there's a permissions issue. Anything interesting in error_7999.log?

**Comment by spiderpug** [comment]
I put some logic into my container to `cat` the log files if Pushpin dies. Guess you were right!

```
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [INFO] (src/mime.c:49) MAX limits.mime_ext_len=128
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [INFO] (src/host.c:74) MAX limits.url_path=256, limits.host_name=256
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [INFO] (src/handler.c:462) MAX limits.handler_stack=102400
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [INFO] (src/config/config.c:86) Loaded handler 1:ipc:///var/run/pushpin/pushpin-m2-out-7999:pushpin-m2-7999:ipc:///var/run/pushpin/pushpin-m2-in-7999:pushpin-m2-7999
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [INFO] (src/config/config.c:228) Loaded route 1:/:handler for host 1:default
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [INFO] (src/config/config.c:324) Loaded 1 hosts for server 1:default_7999
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [INFO] (src/server.c:539) LOADING Handler ipc:///var/run/pushpin/pushpin-m2-out-7999
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [ERROR] (src/unixy.c:158: errno: No such file or directory) Process 9 is still running, shut itdown first.
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [ERROR] (src/mongrel2.c:180: errno: None) Failed to remove the dead PID file: /var/run/pushpin/mongrel2_7999.pid
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [ERROR] (src/mongrel2.c:418: errno: None) PID file failure, aborting rather than trying to start.
pushpin_1  | Wed, 05 Apr 2017 19:25:50 GMT [ERROR] (src/mongrel2.c:450: errno: None) Exiting due to error.
```

I will invest some more time to figure out why the files are still there and which files to clean in particular.

Thanks for the hint to the log files!

**Comment by jorgecuesta** [comment]
Hi! something new about this issue? I had the exactly same trouble with docker/pushpin/osx

**Comment by brunoalano** [comment]
@jkarneges Do you know when it's going to be updated? I had the same issue

**Comment by jkarneges** [comment]
The new mongrel2 was released. However I'm not sure if the issue is truly fixed, as I believe I'm running the latest version here and I've still seen the "exits uncleanly" message a few times since then. You might give it a test yourself though.

If you're running into the related issue where the PID file isn't cleaned up and mongrel2 refuses to start again, that's been fixed in Pushpin's develop branch.

**Comment by brunoalano** [comment]
Ok, using the `develop` branch fixed the issue. Thanks @jkarneges

**Comment by jkarneges** [comment]
Just wanted to note that Pushpin's Mongrel2 pidfile handling fix was released in 1.18.0. This should solve the problem of a stale pidfile causing Pushpin to fail to start.

The latest Mongrel2 may still sometimes exit uncleanly causing the unsightly log message, though. Will keep this issue open until that's fixed.

**Comment by jkarneges** [comment]
The latest version of Pushpin (1.31.0) uses [Condure](https://github.com/fanout/condure) instead of Mongrel2, so we will no longer track this issue.
